arm/Asmexpand.ml

```diff
@@ -11,8 +11,334 @@
+let _4 = coqint_of_camlint 4l
+let _8 = coqint_of_camlint 8l
+let _16 = coqint_of_camlint 16l
+
+(* Emit instruction sequences that set or offset a register by a constant. *)
+(* No S suffix because they are applied to SP most of the time. *)
+
+let expand_movimm dst n =
+  match Asmgen.decompose_int n with
+  | [] -> assert false
+  | hd::tl->
+     emit (Pmov (dst,SOimm hd));
+     List.iter
+       (fun n -> emit (Porr (dst,dst, SOimm n))) tl
+
+let expand_subimm dst src n =
+  match Asmgen.decompose_int n with
+  | [] -> assert false
+  | hd::tl ->
+     emit (Psub(dst,src,SOimm hd));
+     List.iter (fun n -> emit (Psub (dst,dst,SOimm n))) tl
+
+let expand_addimm dst src n =
+  match Asmgen.decompose_int n with
+  | [] -> assert false
+  | hd::tl ->
+     emit (Padd (dst,src,SOimm hd));
+     List.iter (fun n -> emit (Padd (dst,dst,SOimm n))) tl
+
+let expand_int64_arith conflict rl fn =
+  if conflict then
+    begin
+      fn IR14;
+      emit (Pmov (rl,SOreg IR14))
+    end else
+    fn rl
+
+
+(* Built-ins.  They come in two flavors:
+   - annotation statements: take their arguments in registers or stack
+     locations; generate no code;
+   - inlined by the compiler: take their arguments in arbitrary
+     registers.
+*)
+
+(* Handling of annotations *)
+
+let expand_annot_val txt targ args res =
+  emit (Pannot (EF_annot(txt,[targ]), List.map (fun r -> AA_base r) args));
+  match args, res with
+  | [IR src], [IR dst] ->
+     if dst <> src then emit (Pmov (dst,SOreg src))
+  | [FR src], [FR dst] ->
+     if dst <> src then emit (Pfcpyd (dst,src))
+  | _, _ -> assert false
+
+
+(* Handling of memcpy *)
+
+(* The ARM has strict alignment constraints for 2 and 4 byte accesses.
+   8-byte accesses must be 4-aligned. *)
+
+let expand_builtin_memcpy_small sz al src dst =
+  let rec copy ofs sz  =
+    if sz >= 8 && al >= 4 && !Clflags.option_ffpu then begin
+	emit (Pfldd (FR7,src,ofs));
+	emit (Pfstd (FR7,dst,ofs));
+	copy (Int.add ofs _8) (sz - 8)
+      end else if sz >= 4 && al >= 4 then begin
+	emit (Pldr (IR14,src,SOimm ofs));
+	emit (Pstr (IR14,dst,SOimm ofs));
+	copy (Int.add ofs _4) (sz - 4)
+      end else if sz >= 2 && al >= 2 then begin
+	emit (Pldrh (IR14,src,SOimm ofs));
+	emit (Pstrh (IR14,dst,SOimm ofs));
+	copy (Int.add ofs _2) (sz - 2)
+      end else if sz >= 1 then begin
+	emit (Pldrb (IR14,src,SOimm ofs));
+	emit (Pstrb (IR14,dst,SOimm ofs));
+	copy (Int.add ofs _1) (sz - 1)
+      end else
+      () in
+  copy _0 sz
+
+let expand_builtin_memcpy_big sz al src dst =
+  assert (sz >= al);
+  assert (sz mod al = 0);
+  assert (src = IR2);
+  assert (dst = IR3);
+  let (load, store, chunksize) =
+    if al >= 4 then (Pldr (IR12,src,SOimm _4), Pstr (IR12,dst,SOimm _4) , 4)
+    else if al = 2 then (Pldrh (IR12,src,SOimm _2), Pstrh (IR12,dst,SOimm _2), 2)
+    else (Pldrb (IR12,src,SOimm _1), Pstrb (IR12,dst,SOimm _1), 1) in
+  expand_movimm IR14 (coqint_of_camlint (Int32.of_int (sz / chunksize)));
```

Comment:
The original code was using post-indexed addressing, where  the index register is incremented after the memory access.  Pldr* and Pstr* implement offset addressing.  The code generated here does not implement memcpy.